src/main/java/org/opensearch/knn/index/codec/nativeindex/NativeIndexWriter.java

```diff
@@ -68,13 +73,29 @@ public class NativeIndexWriter {
      * @return correct NativeIndexWriter to make index specified in fieldInfo
      */
     public static NativeIndexWriter getWriter(final FieldInfo fieldInfo, SegmentWriteState state) {
-        final KNNEngine knnEngine = extractKNNEngine(fieldInfo);
-        boolean isTemplate = fieldInfo.attributes().containsKey(MODEL_ID);
-        boolean iterative = !isTemplate && KNNEngine.FAISS == knnEngine;
-        if (iterative) {
-            return new NativeIndexWriter(state, fieldInfo, MemOptimizedNativeIndexBuildStrategy.getInstance());
-        }
-        return new NativeIndexWriter(state, fieldInfo, DefaultIndexBuildStrategy.getInstance());
+        return createWriter(fieldInfo, state, null);
+    }
+
+    /**
+     * Gets the correct writer type for the specified field, using a given QuantizationModel.
+     *
+     * This method returns a NativeIndexWriter instance that is tailored to the specific characteristics
+     * of the field described by the provided FieldInfo. It determines whether to use a template-based
+     * writer or an iterative approach based on the engine type and whether the field is associated with a template.
+     *
+     * If quantization is required, the QuantizationModel is passed to the writer to facilitate the quantization process.
+     *
+     * @param fieldInfo          The FieldInfo object containing metadata about the field for which the writer is needed.
+     * @param state              The SegmentWriteState representing the current segment's writing context.
+     * @param quantizationState  The QuantizationState that contains  quantization state required for quantization
+     * @return                   A NativeIndexWriter instance appropriate for the specified field, configured with or without quantization.
+     */
+    public static NativeIndexWriter getWriter(
+        final FieldInfo fieldInfo,
+        final SegmentWriteState state,
+        final QuantizationState quantizationState
+    ) {
+        return createWriter(fieldInfo, state, quantizationState);
     }
```

Comment:
I don't think so you need createWriter function. because all you are doing is calling the that function. actually you can do something like this.

```
public static NativeIndexWriter getWriter(final FieldInfo fieldInfo, SegmentWriteState state) {
    getWriter(fieldInfo, state, null)
}

public static NativeIndexWriter getWriter(
        final FieldInfo fieldInfo,
        final SegmentWriteState state,
        final QuantizationState quantizationState
    ) {
... code from create Writer here.
} 

```

Reply:
It is required so if tomorrow if I need to create another overload , I can easily extend private method , rather changing all public , so it's clean.

Reply:
cannot say that. Its one and the same thing. Leaving it upto your choice